kafka-connect-s3/src/main/java/io/confluent/connect/s3/format/bytearray/ByteArrayRecordWriterProvider.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright 2017 Confluent Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.confluent.connect.s3.format.bytearray;
+
+import io.confluent.connect.s3.S3SinkConnectorConfig;
+import io.confluent.connect.s3.storage.S3OutputStream;
+import io.confluent.connect.s3.storage.S3Storage;
+import io.confluent.connect.storage.format.RecordWriter;
+import io.confluent.connect.storage.format.RecordWriterProvider;
+import org.apache.kafka.connect.converters.ByteArrayConverter;
+import org.apache.kafka.connect.errors.ConnectException;
+import org.apache.kafka.connect.errors.DataException;
+import org.apache.kafka.connect.sink.SinkRecord;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+
+public class ByteArrayRecordWriterProvider implements RecordWriterProvider<S3SinkConnectorConfig> {
+
+  private static final Logger log = LoggerFactory.getLogger(ByteArrayRecordWriterProvider.class);
+  private static final String EXTENSION = ".out";
```

Comment:
I don't know the right approach for extension. `.out` is intended as a placeholder for now.

My use case for this format is that I have a topic with schema-less JSON messages and I want to write them unchanged and unparsed to S3. So Kafka Connect can treat the messages as byte arrays and never know they're JSON, but I'd like the output file to have a ".json" extension.

Has there been any thought about making the extension configurable for formats?

Reply:
I'd use `.bin` as a default. 

And also, I think we enabled the way to define a "user-defined" per formatter config property. Although we want to be prudent and not add too many configuration parameters, maybe it would make sense to add this one only for this formatter (not in general for any formatter).

Reply:
> I think we enabled the way to define a "user-defined" per formatter config property

Can you give me any more context on this? I'm not sure if you're referring to an upcoming change or something that already exists.

Reply:
It looks like it's going to be a bit tortured to apply config values to `ByteArrayRecordWriterProvider` since the `conf` doesn't get passed in to the constructor. I could add a param to the constructor to pass in the conf, but that seems to be getting a bit redundant.

Perhaps again this is due to missing context about how user-defined per formatter configs are imagined to work.